### PR TITLE
Bump up opentelemetry metrics to 1.7

### PR DESCRIPTION
**Pull request by worekleszczy**
There are numerous changes in metrics API (most of the breaking changes happened at 1.5 release - repo has been updated to fulfill new metrics [specification](https://github.com/open-telemetry/opentelemetry-specification)).
Most notable changes:
* no more values recorders - at this point we have histograms 🎉 
* Labels were removed - now we use Attributes
* there is distinction between gauge and upDownCounter
* no more BatchRecorder - we can't do atomic update on several metrics at this point. This will be implemented in the future
* different way to set up MetricsProducers

### CODE REVIEW

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/LabelsFactory.scala [hunk 1]
```diff
@@ -1,48 +1,71 @@
 package io.scalac.mesmer.extension.upstream
 
-import io.opentelemetry.api.metrics.common.Labels
+import io.opentelemetry.api.common.AttributeKey
+import io.opentelemetry.api.common.Attributes
 
 import scala.annotation.switch
-import scala.collection.mutable.ArrayBuffer
 
 object LabelsFactory {
```
Comment (lgajowy) [inline extension/src/main/scala/io/scalac/mesmer/extension/upstream/LabelsFactory.scala:8]: The name should change to `AttributesFactory` imho. It seems like `Labels` is a deprecated concept in opentelemetry specification now (I have to make sure)

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala [hunk 2]
```diff
@@ -3,36 +3,45 @@ package io.scalac.mesmer.extension.upstream.opentelemetry
...
+    )
+
+final class LongSumObserverBuilderAdapter[L <: LabelSerializable](builder: LongCounterBuilder)
+    extends MetricObserverBuilderAdapter[ObservableLongMeasurement, Long, L](
+      result => builder.buildWithCallback(result.asJava),
+      longResultWrapper
+    )
+
+sealed abstract class MetricObserverBuilderAdapter[R <: ObservableMeasurement, T, L <: LabelSerializable](
+//  builder: AsynchronousInstrumentBuilder[R],
```
Comment (lgajowy) [inline extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala:43]: Commented out code - figure out why, then fix/remove.

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala [hunk 3]
```diff
@@ -3,36 +3,45 @@ package io.scalac.mesmer.extension.upstream.opentelemetry
...
+      longResultWrapper
+    )
+
+final class LongSumObserverBuilderAdapter[L <: LabelSerializable](builder: LongCounterBuilder)
+    extends MetricObserverBuilderAdapter[ObservableLongMeasurement, Long, L](
+      result => builder.buildWithCallback(result.asJava),
+      longResultWrapper
+    )
+
+sealed abstract class MetricObserverBuilderAdapter[R <: ObservableMeasurement, T, L <: LabelSerializable](
```
Comment (lgajowy) [inline extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala:42]: Do we even need the L type parameter here? It does not seem to be propagated/used anywhere (unless I'm missing anything)

**Review by lgajowy — CHANGES_REQUESTED** [review]
Small review. I will be taking over this PR so this is more like notes for myself but comments are welcome

**Comment by lgajowy** [comment]
Currently, when starting the example, I'm getting: 

```
[error] Could not load Logmanager "wvlet.log.AirframeLogManager"
[error] java.lang.ClassNotFoundException: wvlet.log.AirframeLogManager
[error]         at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
[error]         at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
[error]         at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
[error]         at java.logging/java.util.logging.LogManager$1.run(LogManager.java:239)
[error]         at java.logging/java.util.logging.LogManager$1.run(LogManager.java:223)
[error]         at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
[error]         at java.logging/java.util.logging.LogManager.<clinit>(LogManager.java:222)
[error]         at java.logging/java.util.logging.Logger.demandLogger(Logger.java:649)
[error]         at java.logging/java.util.logging.Logger.getLogger(Logger.java:718)
[error]         at java.logging/java.util.logging.Logger.getLogger(Logger.java:702)
[error]         at io.grpc.ManagedChannelRegistry.<clinit>(ManagedChannelRegistry.java:40)
[error]         at io.grpc.ManagedChannelProvider.provider(ManagedChannelProvider.java:41)
[error]         at io.grpc.ManagedChannelBuilder.forTarget(ManagedChannelBuilder.java:76)
[error]         at io.opentelemetry.exporter.otlp.internal.grpc.DefaultGrpcExporterBuilder.build(DefaultGrpcExporterBuilder.java:117)
[error]         at io.opentelemetry.exporter.otlp.metrics.OtlpGrpcMetricExporterBuilder.build(OtlpGrpcMetricExporterBuilder.java:128)
[error]         at io.opentelemetry.exporter.otlp.metrics.OtlpGrpcMetricExporter.getDefault(OtlpGrpcMetricExporter.java:30)
[error]         at example.Boot$.initOpenTelemetryMetrics(Boot.scala:47)
[error]         at example.Boot$.startUp(Boot.scala:77)
[error]         at example.Boot$.delayedEndpoint$example$Boot$1(Boot.scala:126)
[error]         at example.Boot$delayedInit$body.apply(Boot.scala:33)
[error]         at scala.Function0.apply$mcV$sp(Function0.scala:39)
[error]         at scala.Function0.apply$mcV$sp$(Function0.scala:39)
[error]         at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
[error]         at scala.App.$anonfun$main$1(App.scala:76)
[error]         at scala.App.$anonfun$main$1$adapted(App.scala:76)
[error]         at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error]         at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error]         at scala.App.main(App.scala:76)
[error]         at scala.App.main$(App.scala:74)
[error]         at example.Boot$.main(Boot.scala:33)
[error]         at example.Boot.main(Boot.scala)
```

Possibly it's just a missing depencency `"org.wvlet.airframe" %% "airframe-log" % "21.12.0"` - to be checked too

**Comment by lgajowy** [comment]
We also need to use `image: otel/opentelemetry-collector-contrib-dev:latest` now in order to be able to use health-check extension in the otel collector. For this one I will create a separate issue.

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/LabelsFactory.scala (hunk 1 above)
Comment (lgajowy) [reply extension/src/main/scala/io/scalac/mesmer/extension/upstream/LabelsFactory.scala]: this commit solves it: 202648586b775bae3ef9ee6f6cd61bf20acb5116

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala (hunk 3 above)
Comment (lgajowy) [reply extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala]: Wrapped metric observer needs it - I somehow missed that. Resolving

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala (hunk 2 above)
Comment (lgajowy) [reply extension/src/main/scala/io/scalac/mesmer/extension/upstream/opentelemetry/WrappedAsynchronousInstrument.scala]: fixed

**Comment by lgajowy** [comment]
https://github.com/ScalaConsultants/mesmer-akka-agent/pull/233#issuecomment-992439070 will be solved separately as is a non blocking issue. Separate ticket: https://github.com/ScalaConsultants/mesmer-akka-agent/issues/241

**Review by lgajowy — APPROVED** [review]

**Review by lgajowy — CHANGES_REQUESTED** [review]
To be tested once again - I may have missed something

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala [hunk 4]
```diff
@@ -58,31 +58,32 @@ final class OpenTelemetryHttpMetricsMonitor(
   import HttpMetricsMonitor._
 
   private lazy val requestTimeRequest = meter
-    .longValueRecorderBuilder(metricNames.requestDuration)
+    .histogramBuilder(metricNames.requestDuration)
```
Comment (lgajowy) [inline extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala:61]: TODO: This causes our grafana example dashboards to fail when calculating max values. Figure out what can we do about this

**Review by lgajowy — CHANGES_REQUESTED** [review]

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala (hunk 4 above)
Comment (lgajowy) [reply extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala:61]: Ok. Now I know what is going on here. We were using the long value recorder as it was the only meter type that allowed us to register all (?) samples as they were, without any aggregation. Thanks to that we were able to calculate: 
 - avg (just take the sum of all response times and divide it by the whole count)
 - max (just take the biggest response time as it is there in the value recorder)
 
In 1.7.0 the `Meter` interface is different and no longer allows creating Value recorders. What we're left with are histograms.  Not that it's not sensible to use them for collecting the response times, it's just that they work a little bit differently. What they track is: 
 - the sum of all observation values (per bucket/total), not the exact sample values
 - the count of all observations (per bucket/total) 
 
 Therefore **we no longer have the ability to get the max value** from this metric as it is the histogram now (how do you get the max value if what you have is the already aggregated "sum of response times"?). The same ofc applies to all former value recorders. This is why we can't see the max plots in Grafana anymore in the example project setup.
 
 
What I propose is: 
 - merge this PR as is. I tested it and was able to see other metrics. I also see the histograms and the avg values seem to be calculated correctly from them
 - have an issue either to: 
   1. provide the max request/response value as a gauge (how feasible is that? Should we do this?)
   2. remove/change the plots in grafana to present something else
   
cc:  @jczuchnowski

**Review by lgajowy — APPROVED** [review]

File: extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala (hunk 4 above)
Comment (jczuchnowski) [reply extension/src/main/scala/io/scalac/mesmer/extension/upstream/OpenTelemetryHttpMetricsMonitor.scala:61]: Ok, sounds reasonable. Let's merge and release, and talk about further improvements.

**Review by jczuchnowski — APPROVED** [review]

**Comment by lgajowy** [comment]
@liosedhel @tumevoiz - the artifacts for 0.5.0 version containing this version bump are in the mvn repo now. Feel free to give it a try! :)

**Comment by liosedhel** [comment]
Thank you @lgajowy, great work! we will give it a try for sure :)

**Comment by tumevoiz** [comment]
Thank you @Igajowy :)

**Pull request merged**